### PR TITLE
Part 4 of the Inception Pipeline series

### PR DESCRIPTION
**Pull request by mechanicalpete**
Climbing trees and look at how to tame the forest!

### CODE REVIEW

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+image: img/inception-pipelines/seed_germination.png
+---
+
+**Part 4 seeding a forest** - A theoretical example for DevOps on AWS
+
+![forest of trees]({{ site.url }}/img/inception-pipelines/part-4-trees-lotsa-trees.png)
+
+## What's The Problem
+
+In this instalment we'll be stepping back (or is it climbing up) and looking at how Inception Pipeline fits into the forest of AWS Accounts that most organisations have. Firstly, I'll cover off what every AWS account needs (regardless of what it's used for). Next, I'll briefly discuss the common accounts that everyone should have. We'll then finish up with a few ideas on combining accounts and pipelines.
```
Comment (Hamo69) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: looking at how *an* Inception Pipeline

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+The ideas discussed in this post also fit alongside the [AWS Landing Zone](https://aws.amazon.com/answers/aws-landing-zone/) solution for managing multi-account AWS environments.
+
+**Note:** When writing this post, I struggled to find a simple, generic, way of what an Inception Pipeline does in each of the use cases below. Terms like product, feature, service, etc just didn't cover all the possible scenarios. The best term I could think up was *aspect*. Hopefully the term *aspect* makes sense (or soon will), so let’s get started!
+
+### What every AWS Account needs
+
+![account and aspect pipelines]({{ site.url }}/img/inception-pipelines/part-4-what-every-account-needs.png)
+
+The above diagram represents a very strong opinion that I have. which is that every AWS account must have at least two Inception Pipelines.
```
Comment (Hamo69) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: *that I have, which*

Comment (Hamo69) [reply _posts/2018-06-27-inception-pipelines-pt4.markdown]: Why must we have 2? Will you explain it later?

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+The first pipeline performs actions and manages the state of the account (aka the account-level pipeline). The obvious items here are things like IAM users/groups/roles/etc, account wide KMS keys, Route 53 hosted zones, etc. The idea here is that anything that changes stuff at the account level is kept together in this pipeline. I would also say it is typically the slower changing items; created once then never really change.
```
Comment (Hamo69) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: _The idea here is that anything that changes_ - *that* used twice in quick succession - remove one

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+The second, and subsequent, pipeline is for doing stuff within the account; they are an aspect of why the account exists in the first place. For example, if the account is for hosting a static website, this is where you put in all the CloudFormation code for configuring CloudFront, S3 buckets, Route 53 domain entries, etc. These items are more likely to have constant change and/or be more transient in lifespan. Another way to think about these aspects is that they are aspects that you want to delete (with the pipeline) but not delete the account.
```
Comment (Hamo69) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: italicise _aspect_ : they are an _aspect_ of

Comment (Hamo69) [reply _posts/2018-06-27-inception-pipelines-pt4.markdown]: _Another way to think about these aspects is that they are aspects that you want to delete (with the pipeline) but not delete the account._ should be 
_Another way to think about these is that they are aspects that you want to delete (with the pipeline) but not delete the account._

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+Pros:
+
+* If non-production is compromised, then production is still safe.
+* Still has the quick feedback cycle on change to non-production environments as where you build the code is also the same AWS account that you run in.
+* Using new AWS services is easier as you don't need to adjust permission in multiple AWS accounts.
+
+Cons:
+
+* When you deploy into Production, it is a different process (i.e. across account boundaries), so the chance for something to go wrong increases.
```
Comment (jgunnink) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: Here you capitalise Production, above (line 129) it's lower cased.

Additionally, in your i.e. bit do you mean _cross_ account boundaries

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+* It can get noisy in the same way that 'All-in-one' does for the non-production environments.
+* Getting the cross-account permissions can be challenging.
+
+#### Environments for Everyone
```
Comment (jgunnink) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: title case inconsistent

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+![cross account components]({{ site.url }}/img/inception-pipelines/part-4-environments-for-everyone.png)
+
+In the final model, we split building the code into its own AWS account, and have one AWS account per environment.
+
+Pros:
+
+* Very clean separation of build vs running code.
+* Environments can progressively get more locked-down/restrictive on who can see/do what.
```
Comment (jgunnink) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: ~get~ become

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+## Where Do I Get The Seed Files
+
+No new files this month. If you're keen to plant an Inception Pipeline in your account, pop across to [GitHub repo and grab the files from there](https://github.com/MechanicalRock/InceptionPipeline/tree/master).
+
+## Taking It For A Spin
+
+If you're unsure where to start, and you thought you could skip reading [Part 1](https://mechanicalrock.github.io/aws/continuous/deployment/2018/03/01/inception-pipelines-pt1), I highly recommend starting there.
+
+I'm also quite susceptible to bribery by coffee, so if you'd like that chat drop me an email at [contact@mechanicalrock.io](mailto:contact@mechanicalrock.io) and ask for **Pipeline Pete**.
```
Comment (jgunnink) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: 😁

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+A perfect use case here is defining IAM users that can assume roles in other accounts. This gives you the following benefits:
+
+1. Centralised management of IAM users. If someone starts or leaves your organisation, you can add or delete them from a single location rather than hunting through all your other AWS accounts.
+2. Centralised management of permissions. You define an IAM Group in the root account that can then assume a role in a child account. So, if the child account defines an Administrator role, you match that with a group here and then add/remove IAM users as required.
+3. All changes to your IAM users and groups are managed as code. This gives you a minimalistic audit trail of who made changes.
+
+Next month I'll expand upon this idea with code snippets. One other consideration too with only creating users in the Root is the [IAM Service Limits](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-limits.html#reference_iam-limits-entities). It's probably not an issue unless you literally have thousands of users, in which case you'll be better off using a Single Sign Identity Provider (because your root account now only contains roles).
```
Comment (jgunnink) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: >One other consideration too with only creating users in the Root is the ...

You may want to revisit this sentence. Suggest to remove the word `too`

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+#### Audit AWS Account
+
+This account is where all things security and audit-ty related go. The account-level Inception Pipeline would define three IAM roles:
+
+* A heavily controlled **Administrator** account. Once defined, it would only need to be used in the case of emergency to perform actions (most likely in the AWS Console itself).
+* A more open role that can be assumed to make changes via the other pipelines. Called something like **SecurityOwner**?
+* Finally, a read-only role that can be given to developers, auditors, etc to view logs and other files stored in this account, and possibly called something like **SecurityReviewer**.
```
Comment (jgunnink) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: remove the comma after account and change `and possibly` to `may be` or `could be`

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+The Sandbox AWS Account is used by developers as a safe place to try out all the services that AWS provides. It is a volatile landscape with lots of things being created and destroyed all the time. Developers should have full organisational and AWS permissions here to experiment. To help prevent cost blowouts, AWS Budgets should be used to alert when the forecasted, and actual, spend hits predefined levels.
+
+It is also possible to define a Lambda in the root account that periodically deletes the Sandbox account and recreates it in a clean state.
+
+The account Inception Pipeline defines the roles that can be assumed, as well as setup AWS Budget alerts. When people are experimenting, they would have their own copy of the pipeline.
+
+### Managing Aspects
+
+Now we move into the interesting bits and answer the question "How do I structure my accounts\pipeline for my application?".
```
Comment (jgunnink) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: some places you use a forward slash others a backslash. Not sure if intentional?

File: _posts/2018-06-27-inception-pipelines-pt4.markdown [hunk 12]
```diff
@@ -0,0 +1,173 @@
+---
+layout: post
+title:  "Seeds of Inception - Part 4"
+date:   2018-06-27
+categories: aws continuous deployment codepipeline codebuild inception pipeline
+author: Pete Yandell
+image: img/inception-pipelines/seed_germination.png
+---
+
+**Part 4 seeding a forest** - A theoretical example for DevOps on AWS
```
Comment (jgunnink) [inline _posts/2018-06-27-inception-pipelines-pt4.markdown]: Title case? Not sure if it matters on actual headings though (is everything capitalised?)

**Pull request merged**